1. Скачайте датасет fifа_s2.сsv. Проанализируйте его и определите, какие данные являются неполными. Удалите ненужные колонки и недостающие значения.

2. Найдите в датафрейме полные дубликаты и удалите их. Значения могут быть одинаковыми, но написаны по-разному. Например, может отличаться размер регистра (заглавные и строчные буквы). Особое внимание уделить колонке с названиями команд.

3. Напишите функцию, которая добавит колонку с разбиением возраста по группам: до 20, от 20 до 30, от 30 до 36 и старше 36. Посчитайте количество футболистов в каждой категории.

In [13]:
import init_spark_env # Ипортирую файл настроек связки спарка и jpyter

In [14]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, count, when, udf, lower
from pyspark.sql.types import StringType

In [15]:
# Запускаю спарк-сессию
spark = SparkSession.builder.appName("hw2 fifa").getOrCreate()

In [16]:
# Проеверяю статус спарка
spark

In [17]:
# Загружаю датасет
df = spark.read.csv("fifa_s2.csv", header=True, inferSchema=True)

In [22]:
df

DataFrame[ID: int, Name: string, Age: int, Nationality: string, Overall: int, Potential: int, Club: string, Value: int, Wage: int, Preferred Foot: string, International Reputation: int, Skill Moves: int, Position: string, Joined: int, Contract Valid Until: string, Height: double, Weight: double, Release Clause: double]

In [21]:
df.head(5)

[Row(ID=1179, Name='G. Buffon', Age=40, Nationality='Italy', Overall=88, Potential=88, Club='Paris Saint-Germain', Value=4000, Wage=77, Preferred Foot='Right', International Reputation=4, Skill Moves=1, Position='GK', Joined=2018, Contract Valid Until='01/01/2019', Height=6.333333333, Weight=203.0, Release Clause=7400.0),
 Row(ID=5479, Name='Casillas', Age=37, Nationality='Spain', Overall=82, Potential=82, Club='FC Porto', Value=1500, Wage=10, Preferred Foot='Left', International Reputation=4, Skill Moves=1, Position=None, Joined=2015, Contract Valid Until='01/01/2019', Height=6.083333333, Weight=185.0, Release Clause=3000.0),
 Row(ID=8205, Name='Braga', Age=35, Nationality='Portugal', Overall=70, Potential=70, Club='CD Aves', Value=750, Wage=5, Preferred Foot='Right', International Reputation=1, Skill Moves=3, Position='LCM', Joined=2017, Contract Valid Until='01/01/2019', Height=5.916666667, Weight=165.0, Release Clause=1600.0),
 Row(ID=9014, Name='A. Robben', Age=34, Nationality='Ne

In [23]:
# Схема датафрейма
df.printSchema()

root
 |-- ID: integer (nullable = true)
 |-- Name: string (nullable = true)
 |-- Age: integer (nullable = true)
 |-- Nationality: string (nullable = true)
 |-- Overall: integer (nullable = true)
 |-- Potential: integer (nullable = true)
 |-- Club: string (nullable = true)
 |-- Value: integer (nullable = true)
 |-- Wage: integer (nullable = true)
 |-- Preferred Foot: string (nullable = true)
 |-- International Reputation: integer (nullable = true)
 |-- Skill Moves: integer (nullable = true)
 |-- Position: string (nullable = true)
 |-- Joined: integer (nullable = true)
 |-- Contract Valid Until: string (nullable = true)
 |-- Height: double (nullable = true)
 |-- Weight: double (nullable = true)
 |-- Release Clause: double (nullable = true)



In [24]:
# Табличный вид
df.show()

+-----+-----------------+---+-------------------+-------+---------+-------------------+-----+----+--------------+------------------------+-----------+--------+------+--------------------+-----------+------+--------------+
|   ID|             Name|Age|        Nationality|Overall|Potential|               Club|Value|Wage|Preferred Foot|International Reputation|Skill Moves|Position|Joined|Contract Valid Until|     Height|Weight|Release Clause|
+-----+-----------------+---+-------------------+-------+---------+-------------------+-----+----+--------------+------------------------+-----------+--------+------+--------------------+-----------+------+--------------+
| 1179|        G. Buffon| 40|              Italy|     88|       88|Paris Saint-Germain| 4000|  77|         Right|                       4|          1|      GK|  2018|          01/01/2019|6.333333333| 203.0|        7400.0|
| 5479|         Casillas| 37|              Spain|     82|       82|           FC Porto| 1500|  10|          Left

In [25]:
# Количество строк
df.count()

2399

In [26]:
# Количество пропущенных значений
df_missing_values = df.select([count(when(col(c).isNull(), c)).alias(c) for c in df.columns])

In [27]:
# Пропуски по колонкам
df_missing_values.show()

+---+----+---+-----------+-------+---------+----+-----+----+--------------+------------------------+-----------+--------+------+--------------------+------+------+--------------+
| ID|Name|Age|Nationality|Overall|Potential|Club|Value|Wage|Preferred Foot|International Reputation|Skill Moves|Position|Joined|Contract Valid Until|Height|Weight|Release Clause|
+---+----+---+-----------+-------+---------+----+-----+----+--------------+------------------------+-----------+--------+------+--------------------+------+------+--------------+
|  0|   0|  0|          7|      0|        0| 241|  257|   0|             0|                      10|          6|      21|     0|                 273|     0|     0|             3|
+---+----+---+-----------+-------+---------+----+-----+----+--------------+------------------------+-----------+--------+------+--------------------+------+------+--------------+



из 2399 максимально с пропусками :
Club - 241, Value -257, Contract Valid Until - 273   - вердикт "удаляю"
Остальные маленькие пропуски по колонкам  -удалю строки

In [28]:
# Удаляю колоноки 
columns_to_drop = ['Club', 'Contract Valid Until', 'Value']
df_cleaned = df.drop(*columns_to_drop)

In [29]:
# Удаляю строки с проусками
df_cleaned = df_cleaned.dropna()

In [30]:
# Все строковые колонки привожу к нижнему регистру для удаления дубликатов
string_columns = [field.name for field in df_cleaned.schema.fields if isinstance(field.dataType, StringType)]
for col_name in string_columns:
    df_cleaned = df_cleaned.withColumn(col_name, lower(col(col_name)))

In [34]:
# Удаляю дубликаты
df_no_duplicates = df_cleaned.dropDuplicates()

In [35]:
# Функция для определения возрастной группы
def age_group(age):
    if age < 20:
        return 'Under 20'
    elif 20 <= age <= 30:
        return '20 to 30'
    elif 31 <= age <= 36:
        return '31 to 36'
    else:
        return 'Over 36'

# Регистрация UDF
age_group_udf = udf(age_group, StringType())

# Добавление новой колонки 'AgeGroup'
df_with_age_groups = df_no_duplicates.withColumn('AgeGroup', age_group_udf(col('Age')))

# Подсчет количества футболистов в каждой возрастной группе
age_group_counts = df_with_age_groups.groupBy('AgeGroup').count()

# Вывод результатов
age_group_counts.show()

+--------+-----+
|AgeGroup|count|
+--------+-----+
|31 to 36|  327|
|Under 20|  269|
| Over 36|   14|
|20 to 30| 1744|
+--------+-----+



In [36]:
spark.stop()